### PR TITLE
Add cattrs to unstructure/structure and allow (de)serialization with json or similar

### PR DESCRIPTION
**Pull request by anthrotype**
this allows to do things like

```python
import orjson
from ufoLib2.converters import default_converter as c
from ufoLib2.objects import Font

font = Font.open("NotoSans-Regular.ufo")

# dumping a UFO to json
data = orjson.dumps(c.unstructure(font))
with open("notosans-regular.json", "wb") as f:
    f.write(data)

# or loading a UFO from json
with open("notosans-regular.json", "rb") as f:
    data = f.read()
font2 = c.structure(orjson.loads(data), Font)

assert font == font2
```

~~Still WIP, as I need to add tests.~~ Added tests

### CODE REVIEW

File: src/ufoLib2/objects/image.py [hunk 1]
```diff
@@ -46,15 +46,15 @@ def __bool__(self) -> bool:
     # the fontTools.ufoLib.validators.imageValidator requires that image is a
     # subclass of Mapping...
 
-    _transformation_keys_: tuple[str, str, str, str, str, str] = (
+    _transformation_keys_: ClassVar[tuple[str, str, str, str, str, str]] = (
```
Comment (madig) [inline src/ufoLib2/objects/image.py:49]: Should we move this stuff to outside the class?

File: src/ufoLib2/objects/lib.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+#     int,
+#     str,
+#     Mapping[str, PlistEncodable],
+#     Sequence[PlistEncodable],
+# ]
+
+
+DEFAULT_ENCODING = "base85"
+
+BYTES_ENCODINGS: dict[
```
Comment (madig) [inline src/ufoLib2/objects/lib.py:31]:
```suggestion
BYTES_ENCODINGS: Dict[
```

File: src/ufoLib2/objects/image.py (hunk 1 above)
Comment (anthrotype) [reply src/ufoLib2/objects/image.py:49]: well, it is used only by the Image class so it is class-specific. I'll keep ClassVar.

File: src/ufoLib2/objects/lib.py (hunk 2 above)
Comment (anthrotype) [reply src/ufoLib2/objects/lib.py:31]: I tried to use the fancy generic types for the postponed annotations that are only used for static typechecking (like in this line), and only use the old-style generic collections with uppercase letters for the places where we need those types at runtime (for cattrs structuring/destructuring or when are used as base class).

**Comment by anthrotype** [comment]
hm gotta figure out py37 compat (next week)

**Comment by anthrotype** [comment]
@madig I think this is ready for another review pass

**Comment by anthrotype** [comment]
By the way, all this is generic and not tied to json or any specific serialization format. Here's how it'd work with msgpack for example. The only difference below is, instead of using ufoLib2's default_converter, we create a new GenConverter, register our custom structure/unstructure hooks, and allow_bytes=True (for msgpack can encode those natively, unlike the our default_converter which is geared toward json and we need to encode bytes as base64).

```python
import io
import cattr
import ufoLib2
import ufoLib2.converters
import msgpack

conv = cattr.GenConverter(omit_if_default=True)
ufoLib2.converters.register_hooks(conv, allow_bytes=True)

font = ufoLib2.Font.open("...ufo")

buf = io.BytesIO()

msgpack.pack(conv.unstructure(font), buf)

buf.seek(0)

font2 = conv.structure(msgpack.unpack(buf), Font)

assert font == font2
```

**Comment by anthrotype** [comment]
This PR also adds support for the missing WOFF metadata attributes in fontinfo.plist, which should fix https://github.com/fonttools/ufoLib2/issues/19

File: src/ufoLib2/objects/misc.py [hunk 3]
```diff
@@ -333,22 +334,42 @@ def _structure(
...
 
-    For use with Anchors and Guidelines classes, where client code
+    For use with Anchors, Guidelines and WoffMetadata classes, where client code
     expects them to behave as dict.
     """
 
     # XXX: Use generics?
 
+    @classmethod
+    @lru_cache(None)
```
Comment (madig) [inline src/ufoLib2/objects/misc.py:347]: `@lru_cache(maxsize=None)`?

Comment (anthrotype) [reply src/ufoLib2/objects/misc.py]: Done

**Comment by madig** [comment]
The LayerSet does not seem to be unstructured?

```
import json

import ufoLib2
import ufoLib2.converters

u = ufoLib2.Font.open("tests/data/MutatorSansBoldCondensed.ufo/")
d = ufoLib2.converters.default_converter.unstructure(u)

with open(r"/tmp/a.json", "w+") as f:
    json.dump(d, f)
```

> TypeError: Object of type LayerSet is not JSON serializable

**Comment by anthrotype** [comment]
> The LayerSet does not seem to get unstructured?

that is very weird, the tests pass, and I added more that explicitly test json (de)serialization. It must be a local setup issue of yours.

**Comment by madig** [comment]
How should other libraries handle serialized dumps? Consider e.g. intermediate build steps in a pipeline, where currently a Designspace might reference normal UFOs but should reference JSON dumps instead. Should e.g. designspaceLib be made aware of the new format or should this rather be done elsewhere?

**Comment by anthrotype** [comment]
we are not designing any "new format" here. This is not tied to json particularly either. Clients of ufoLib2 can do whatever they like with this, this is just enabling them to more easily (de)construct ufoLib2 classes to something that can be (de)serialized. We can figure those thing out elsewhere

**Comment by madig** [comment]
Right, so we use it externally to the established libraries, where tooling has to deliberately make use of it.

**Pull request merged**